src/roompanel/RoomManager.cpp

```diff
@@ -0,0 +1,252 @@
+    if (!doc.isObject()) {
+        if (false) {
+            qWarning().noquote() << "RoomManager received GMCP" << msg.getName().toQString()
+                                 << "containing invalid Json: expecting object, got"
+                                 << msg.getJson().toQString();
+        }
+        return;
+    }
+    updateMob(doc.object());
+}
+
+void RoomManager::addMob(QJsonObject obj)
+{
+    RoomMobUpdate data;
+    if (toMob(obj, data)) {
+        m_room->addMob(std::move(data));
+        updateWidget();
+    }
+}
+
+void RoomManager::updateMob(QJsonObject obj)
+{
+    RoomMobUpdate data;
+    if (toMob(obj, data)) {
+        if (m_room->updateMob(std::move(data))) {
+            updateWidget();
+        }
+    }
+}
+
+bool RoomManager::isInvalidJson(const GmcpMessage &msg)
+{
+    if (msg.hasError()) {
+        if (false) {
+            qWarning().noquote() << "RoomManager received GMCP containing invalid Json:"
+                                 << msg.getError().errorString();
+        }
+        return true;
+    }
+    return false;
+}
+
+void RoomManager::showGmcp(const GmcpMessage &msg)
+{
+    if (false) {
+        qDebug().noquote() << "RoomManager received GMCP:" << msg.toRawBytes();
+    }
+}
+
+const QHash<QString, MobFieldEnum> RoomManager::mobFields{
+    {"name", MobFieldEnum::NAME},
+    {"desc", MobFieldEnum::DESC},
+    {"fighting", MobFieldEnum::FIGHTING},
+    {"flags", MobFieldEnum::FLAGS},
+    {"labels", MobFieldEnum::LABELS},
+    {"riding", MobFieldEnum::MOUNT},
+    {"driving", MobFieldEnum::MOUNT},
+    {"position", MobFieldEnum::POSITION},
+    {"weapon", MobFieldEnum::WEAPON},
+};
+
+bool RoomManager::toMob(QJsonObject obj, RoomMobUpdate &data)
+{
+    QJsonObject::iterator iter = obj.find("id");
+    QJsonObject::iterator end = obj.end();
+    if (iter == end || !toMobId(*iter, data.id())) {
+        return false;
+    }
+    auto matchEnd = mobFields.end();
+    for (iter = obj.begin(); iter != end; ++iter) {
+        auto match = mobFields.find(iter.key());
+        if (match != matchEnd) {
+            toMobField(*iter, data, *match);
+        }
+    }
+    return true;
+}
+
+bool RoomManager::toMobId(QJsonValueRef value, RoomMob::Id &id)
+{
+    // RoomMob::Id cannot represent negative numbers
+    const double d = value.toDouble(-1.0);
+    id = static_cast<RoomMob::Id>(d);
+    // check for exact conversion
+    return std::equal_to<double>{}(d, static_cast<double>(id));
+}
+
+void RoomManager::toMobField(QJsonValueRef value, RoomMobUpdate &data, MobFieldEnum i)
+{
+    QString str;
+    QVariant &variant = data.field(i);
+
+    switch (value.type()) {
+    case QJsonValue::Double:
+        variant = static_cast<unsigned>(value.toDouble());
+        break;
+    case QJsonValue::String:
+        str = value.toString();
+        if (str == RoomMobData::you) {
+            str = RoomMobData::YOU;
+        }
+        if (!str.isEmpty() && str.at(0) == QChar('*')) {
```

Comment:
Please add a comment for why `*`. I imagine this is for `*an Orc*` but it's not super clear.

Reply:
Exactly, it's to highlight enemy players. I added a comment.